✨ FYSETC Spider King V1.1 (STM32H723ZG)

zzcatvs opened the pull request:
Submitting a Pull Request
Based on the FYSETC SPIDER KING V1.1 mainboard that was merged into Marlin, this is for the FYSETC Spider King v1.1 (STM32F723ZGT6).


### Description
Based on the FYSETC SPIDER KING V1.1 mainboard that was merged into Marlin, this is for the FYSETC Spider King v1.1 (STM32F723ZGT6).

### Requirements

Based on the FYSETC SPIDER KING V1.1 mainboard that was merged into Marlin, this is for the FYSETC Spider King v1.1 (STM32F723ZGT6).


### Related Issues

add FYSETC SPIDER KING V1.1 board.

Commit:
add FYSETC SPIDER KING V1.1 board

thisiskeithb commented:
> STM32F723ZGT6

I think you mean STM32H723ZGT6

Also, you’ll need to revert `Configuration.h` & `platformio.ini` changes.

thisiskeithb committed:
Revert Configuration.h changes

thisiskeithb committed:
STM32F723ZGT6 => STM32H723ZGT6

thisiskeithb committed:
Revert platformio.ini changes

thisiskeithb committed:
Format pins_FYSETC_SPIDER_KING723.h, add new line in boards.h

thisiskeithb committed:
Shorten MCU type

thisiskeithb committed:
(c) 2024

thisiskeithb committed:
env:KING723 => env:FYSETC_SPIDER_KING723

...to match FYSETC_SPIDER_KING407 formatting

thisiskeithb committed:
Update comment spacing

thisiskeithb committed:
Add PROBE_ENABLE_PIN

thisiskeithb committed:
Rework onboard NeoPixel conditionals

thisiskeithb committed:
Use ANY() macro

thisiskeithb commented on Marlin/src/pins/stm32h7/pins_FYSETC_SPIDER_KING723.h:
```diff
@@ -0,0 +1,417 @@
+  #elif !defined(Z_MIN_PROBE_PIN)
+    #define Z_MIN_PROBE_PIN                 PA0   // PD0
+  #endif
+#endif
```
This seems more complicated than it needs to be..

zzcatvs replied:
Sorry, I didn't do it well enough. I can make it simpler. However, as long as it doesn't affect usage, there may be some bugs when copying and pasting.

thisiskeithb commented on Marlin/src/pins/stm32h7/pins_FYSETC_SPIDER_KING723.h:
```diff
@@ -0,0 +1,417 @@
+//
+// Servos
+//
+#define SERVO0_PIN                          PA1   // PD1
```
Do you know why this (and other pins) have comments of other pins? I can't find any information on this variant of the Spider King to confirm what the silkscreen labeling says.

zzcatvs replied:
You can check here for more information: https://wiki.fysetc.com/FYSETC-SPIDER-KING
The pin definition is exactly the same, and the new STM32H723 chip was used after replacing the motherboard.

thisiskeithb replied:
Ahh. I didn't realize it was mostly the same as [`pins_FYSETC_SPIDER_KING407.h`](https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.1.x/Marlin/src/pins/stm32f4/pins_FYSETC_SPIDER_KING407.h).

This can be simplified a lot further & combined, but I'll have to do that once I'm back at a computer.

zzcatvs commented:
Thank you.

thisiskeithb committed:
Use *_common.h pin scheme

thisiskeithb commented:
> The pin definition is exactly the same, and the new STM32H723 chip was used after replacing the motherboard.

I've updated this PR to use our `*_common.h` pin scheme since the only difference is the processor. The H723 version will use the I2C EEPROM as originally defined for the F407 version instead of an emulated EEPROM.

Also, the ability to upgrade the MCU via daughterboard sure is neat!

thisiskeithb commented on ini/stm32h7.ini:
```diff
@@ -137,3 +137,29 @@ board                       = marlin_STM32H723ZE
+                              -DD_CACHE_DISABLED
+                              -UI2C5_BASE
+upload_protocol             = cmsis-dap
+debug_tool                  = cmsis-dap
```
I just noticed that this is nearly an exact copy of `[STM32H723Zx_btt]` / BigTreeTech STM32H723Zx common environment.

Are all these build flags applicable? Was this PR tested on real hardware?

zzcatvs replied:
Thank you for your excellent work, which has been of great help to me. Just now, I verified the modified code.

A. After modification, the code:```Board-INFO-NAME" Spider King V1.1```may not be displayed fully under 12864, so I suggest changing it to``` Board-INFO-NAME "SpiderKing V1.1``` to display all content normally.

B. There seems to be an issue with using I2C EEPROM. After enabling EEPROM, I used m501 and m500:```echo: EEPROM version mismatch (EEPROM=?"? "Marlin=V90)```, add the following code under pins-FYSETC-SPIDER_KING723. h: 

```
#if ANY(NO_EEPROM_SELECTED, FLASH_EEPROM_EMULATION)
  #undef NO_EEPROM_SELECTED
  #ifndef FLASH_EEPROM_EMULATION
    #define FLASH_EEPROM_EMULATION
  #endif
  #define EEPROM_PAGE_SIZE      (0x800UL) // 2K
  #define EEPROM_START_ADDRESS  (0x8000000UL + (STM32_FLASH_SIZE) * 1024UL - (EEPROM_PAGE_SIZE) * 2UL)
  #define MARLIN_EEPROM_SIZE    EEPROM_PAGE_SIZE
#endif
```

can read and write eproom normally, The issue with i2c may require your assistance.

C. "[STM32H723Zx_btt] is similar. In fact, STM32H723VGT6 is also similar to this configuration and can be reused. However, because it is STM32H723Zx_btt, I cannot use it directly. Perhaps I can change it:

```
[STM32H723]
extends                     = stm32_variant
platform                    =  ststm32@15.4.1
platform_packages           = framework-arduinoststm32@~4.20200.220530
board_build.offset          = 0x20000
board_upload.offset_address = 0x08020000
Build_flags=${stm32_variant. build_flags}
-DPIN-SERIAL1TX=PA_9- DPIN-SERIAL1RX=PA_10
-DPIN-SERIAL2TX=PD_5- DPIN-SERIAL2RX=PD_6
-DPIN-SERIAL3-TX=PD_8- DPIN-SERIAL3-RX=PD_9
-DPIN-SERIAL4-TX=PA-0- DPIN-SERIAL4-RX=PA_1
-DPIN-SERIAL7-TX=PE_8- DPIN-SERIAL7-RX=PE_7
-DSERIAL-RX-BUFFER-SIZE=1024- DSERIAL-TX-BUFFER-SIZE=1024
-DTIMER_SERVO=TIM5- DTIMER-TONE=TIM2
-DSTEP_TIMER-IRQ-PRIO=0
-DRCC-PERIPHCLKI2C35=RCCPERIPHCLKI2C5
-DUSE-USB-HS - DUSE-USB-HS.IN-FS
-DD-CACHE-DISABLED
-UI2C5-BASE
upload_protocol             = cmsis-dap
debug_tool                  = cmsis-dap

[env: FYSETL_SPIDER_KING723]
extends                     = STM32H723
board                       = marlin_STM32H723ZG

[env: FYSETL_SPIDER_30]
extends                     = STM32H723
board                       = marlin_STM32H723VG

......
```


It would be better, of course, this is just a suggestion.
![微信图片_20240527162523](https://github.com/MarlinFirmware/Marlin/assets/145735142/97c2b875-dd28-4925-8262-1d4b1fe37682)

thisiskeithb replied:
> After modification, the code:`Board-INFO-NAME" Spider King V1.1`may not be displayed fully under 12864, so I suggest changing it to` Board-INFO-NAME "SpiderKing V1.1` to display all content normally.

> <img width="483" alt="image" src="https://github.com/MarlinFirmware/Marlin/assets/13375512/be458c13-48f7-4bd4-857b-7b428a07e7ff">

Since you didn't set a `CUSTOM_MACHINE_NAME`, Marlin is defaulting it to `DEFAULT_MACHINE_NAME` which defaults to `BOARD_INFO_NAME` when not defined.

As long as `BOARD_INFO_NAME` is 20 characters or less, that'll work fine under the "About Printer" menu, but you should be setting your own `CUSTOM_MACHINE_NAME` so it shows up correctly on the Status screen.

> There seems to be an issue with using I2C EEPROM. After enabling EEPROM, I used m501 and m500:echo: EEPROM version mismatch (EEPROM=?"? "Marlin=V90)

Hmm. Maybe FYSETC removed the external EEPROM in later Spider King versions? @GerogeFu?

zzcatvs replied:
Thank you for your help. Yes, just make the necessary changes to enable it：

```
// Name displayed in the LCD "Ready" message and Info menu
#define CUSTOM_MACHINE_NAME "3D Printer"
```

will not be displayed when defining CUSTOM-Machine-NAME, but when I haven't set it up, I also hope it can display the complete motherboard information. Okay, I'm a bit obsessive-compulsive.

> Hmm. Maybe FYSETC removed the external EEPROM in later Spider King versions?

FYSETC did not remove Spider King external eproom, which can be used under the F4 chip, but I cannot use it after replacing it with the H7 chip. Perhaps there are other reasons.

I'm not GerogeFu.

thisiskeithb replied:
> FYSETC did not remove Spider King external eproom, which can be used under the F4 chip, but I cannot use it after replacing it with the H7 chip

This is where things can get complicated since you can technically have a Spider King V1.0 or V1.1 with STM32F4 or STM32H7 MCUs. This means a working EEPROM implementation isn't tied to the specific board, but board + MCU combination.

I'll have to think of how that fits into Marlin's way of handling board versions now that we have that MCU variable.

Edit: See 4145eae + 0de6848 + d7d3766

> I'm not GerogeFu.

They are the FYSETC point of contact and have contributed to the project, so I was hoping they would have more information.

thisiskeithb committed:
STM32H723Zx_btt => STM32H723Zx_common, extend for Spider King H723

thisiskeithb committed:
Restore emulated EEPROM for Spider King H723

thisiskeithb committed:
Move I2C EEPROM defines to Spider King F407 pins

This can be reverted if we hear back from FYSETC about I2C EEPROM usage on newer boards.

thisiskeithb committed:
Move EEPROM defines back to common pins & use NOT_TARGET() macro

thisiskeithb committed:
Flip EEPROM conditional

thisiskeithb committed:
Ensure Spider King V1.0 & V1.1 exist for both STM32F4 & STM32H7

thisiskeithb committed:
env:FYSETC_SPIDER_KING723 => env:STM32H723ZG_fysetc

thisiskeithb committed:
env:FYSETC_SPIDER_KING407 => env:STM32F407ZG_fysetc

thisiskeithb committed:
Fix up boards.h & pins.h

thisiskeithb commented:
I've rebased this PR to bring it up to date with current `bugfix-2.1.x` & resolved merge conflicts, so please update your local copy with the following before pushing additional commits:

```shell
git checkout FYSETC-SPIDER-KING-H7
git fetch origin
git reset --hard origin/FYSETC-SPIDER-KING-H7
```

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/27125

thinkyhead committed:
merge followup
